meter/sma/listener.go

```diff
@@ -0,0 +1,225 @@
+	"1:30.4.0": {4, 0.1}, "1:30.8.0": {8, 1}, // Scheinleistung (W)/-arbeit (Ws) −
+	"1:31.4.0": {4, 0.001}, // Strom (A)
+	"1:32.4.0": {4, 0.001}, // Spannung (V
+	// Phase 2: {
+	"1:41.4.0": {4, 0.1}, "1:41.8.0": {8, 1}, // Wirkleistung (W)/-arbeit (Ws) +
+	"1:42.4.0": {4, 0.1}, "1:42.8.0": {8, 1}, // Wirkleistung (W)/-arbeit (Ws) −
+	"1:43.4.0": {4, 0.1}, "1:43.8.0": {8, 1}, // Blindleistung (W)/-arbeit (Ws) +
+	"1:44.4.0": {4, 0.1}, "1:44.8.0": {8, 1}, // Blindleistung (W)/-arbeit (Ws) −
+	"1:49.4.0": {4, 0.1}, "1:49.8.0": {8, 1}, // Scheinleistung (W)/-arbeit (Ws) +
+	"1:50.4.0": {4, 0.1}, "1:50.8.0": {8, 1}, // Scheinleistung (W)/-arbeit (Ws) −
+	"1:51.4.0": {4, 0.001}, // Strom (A)
+	"1:52.4.0": {4, 0.001}, // Spannung (V)
+	// Phase 3: {
+	"1:61.4.0": {4, 0.1}, "1:61.8.0": {8, 1}, // Wirkleistung (W)/-arbeit (Ws) +
+	"1:62.4.0": {4, 0.1}, "1:62.8.0": {8, 1}, // Wirkleistung (W)/-arbeit (Ws) −
+	"1:63.4.0": {4, 0.1}, "1:63.8.0": {8, 1}, // Blindleistung (W)/-arbeit (Ws) +
+	"1:64.4.0": {4, 0.1}, "1:64.8.0": {8, 1}, // Blindleistung (W)/-arbeit (Ws) −
+	"1:69.4.0": {4, 0.1}, "1:69.8.0": {8, 1}, // Scheinleistung (W)/-arbeit (Ws) +
+	"1:70.4.0": {4, 0.1}, "1:70.8.0": {8, 1}, // Scheinleistung (W)/-arbeit (Ws) −
+	"1:71.4.0": {4, 0.001}, // Strom (A)
+	"1:72.4.0": {4, 0.001}, // Spannung (V)
+	// Others
+	"144:0.0.0": {4, 1}, // SW Version
+}
+
+var Instance *Listener
+
+// TelegramData defines the data structure of a SMA multicast data package
+type TelegramData struct {
+	Addr   string
+	Serial string
+	Data   map[string]float64
+}
+
+// Listener for receiving SMA multicast data packages
+type Listener struct {
+	mux     sync.Mutex
+	log     *api.Logger
+	conn    *net.UDPConn
+	clients map[string]chan<- TelegramData
+}
+
+// New creates a Listener
+func New(log *api.Logger, addr string) *Listener {
+	// Parse the string address
+	laddr, err := net.ResolveUDPAddr("udp4", multicastAddr)
+	if err != nil {
+		log.FATAL.Fatalf("error resolving udp address: %s", err)
+	}
+
+	// Open up a connection
+	conn, err := net.ListenMulticastUDP("udp4", nil, laddr)
+	if err != nil {
+		log.FATAL.Fatalf("error opening connecting: %s", err)
+	}
+
+	if err := conn.SetReadBuffer(udpBufferSize); err != nil {
+		log.FATAL.Fatalf("error setting read buffer: %s", err)
+	}
+
+	l := &Listener{
+		log:  log,
+		conn: conn,
+	}
+
+	go l.listen()
+
+	return l
+}
+
+// processUDPData converts a SMA Multicast data package into TelegramData
+func (l *Listener) processUDPData(src *net.UDPAddr, buffer []byte) (TelegramData, error) {
+	numBytes := len(buffer)
+
+	if numBytes < 29 {
+		return TelegramData{}, errors.New("received data package is too small")
+	}
+
+	serial := strconv.FormatUint(uint64(binary.BigEndian.Uint32(buffer[20:24])), 10)
+
+	obisCodeValues := make(map[string]float64)
+
+	// read obis code values, start at position 28, after initial static stuff
+	for i := 28; i < numBytes; i++ {
```

Comment:
Suggesting to move this into a new ticket. A bit tired to transform this right now. I'll setup a ticket. Since we have test cases, it should be possible to do without the device